File: 85.最大矩形.go
/*
 * @lc app=leetcode.cn id=85 lang=golang
 *
 * [85] 最大矩形
 */

// @lc code=start
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maximalRectangle(matrix [][]byte) int {
	m := len(matrix)
	if m == 0 {
		return 0
	}
	n := len(matrix[0])
	if n == 0 {
		return 0
	}

	left := make([]int, n)
	right := make([]int, n)
	for k := 0; k < n; k++ {
		right[k] = n
	}
	res := 0
	heights := make([]int, n)
	for i := 0; i < m; i++ {
		curLeft := 0
		for j := 0; j < n; j++ {
			if matrix[i][j] == '1' {
				heights[j]++
				left[j] = max(left[j], curLeft)
			} else {
				heights[j] = 0
				left[j] = 0 //
				curLeft = j + 1
			}
		}

		//fmt.Println("left:", left)
		curRight := n
		for j := n - 1; j >= 0; j-- {
			if matrix[i][j] == '1' {
				right[j] = min(right[j], curRight)
			} else {
				right[j] = n //
				curRight = j
			}
		}
		//fmt.Println("right", right)

		//fmt.Println("height", heights)
		for j := 0; j < n; j++ {
			res = max(res, heights[j]*(right[j]-left[j]))
		}
	}
	return res
}

// @lc code=end

File: 85.最大矩形.py
#
# @lc app=leetcode.cn id=85 lang=python3
#
# [85] 最大矩形
#
# https://leetcode-cn.com/problems/maximal-rectangle/description/
#
# algorithms
# Hard (50.16%)
# Likes:    734
# Dislikes: 0
# Total Accepted:    55.6K
# Total Submissions: 110.9K
# Testcase Example:  '[["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]'
#
# 给定一个仅包含 0 和 1 、大小为 rows x cols 的二维二进制矩阵，找出只包含 1 的最大矩形，并返回其面积。
# 
# 
# 
# 示例 1：
# 
# 
# 输入：matrix =
# [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
# 输出：6
# 解释：最大矩形如上图所示。
# 
# 
# 示例 2：
# 
# 
# 输入：matrix = []
# 输出：0
# 
# 
# 示例 3：
# 
# 
# 输入：matrix = [["0"]]
# 输出：0
# 
# 
# 示例 4：
# 
# 
# 输入：matrix = [["1"]]
# 输出：1
# 
# 
# 示例 5：
# 
# 
# 输入：matrix = [["0","0"]]
# 输出：0
# 
# 
# 
# 
# 提示：
# 
# 
# rows == matrix.length
# cols == matrix[0].length
# 0 
# matrix[i][j] 为 '0' 或 '1'
# 
# 
#

# @lc code=start
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        m = len(matrix)
        if m == 0:
            return 0
        n = len(matrix[0])
        if n == 0:
            return 0
        left = [0] * n
        right = [n] * n
        height = [0] * n
        res = 0
        for i in range(m):
            curLeft = 0
            for j in range(n):
                if matrix[i][j] == '1':
                    height[j] += 1
                    left[j] = max(curLeft, left[j])
                else:
                    height[j] = 0
                    left[j] = 0
                    curLeft = j + 1
            
            curRight = n
            for j in range(n - 1, -1, -1):
                if matrix[i][j] == '1':
                    right[j] = min(curRight, right[j])
                else:
                    right[j] = n
                    curRight = j
            
            for j in range(n):
                res = max(res, height[j]*(right[j] - left[j]))
        return res
                

# @lc code=end

